### ISSUE TITLE
Enable AWS Shield mitigation using Config auto-remediation

### ISSUE THREAD

**Issue by patheard** [cds-snc/scan-files#486]
# Summary
AWS Shield offers auto-mitigation of DDoS attacks.  However, enabling this causes the Terraform definition for our WAF ACLs to get out of sync with the infrastructure since the auto-mitigation adds a new AWS rule to the ACL.

There is an active AWS provider issue tracking the problem, but no official fix yet.

However, there is a Config based solution that may accomplish what we want:
1. Continue to use Terraform to manage our WAF ACLs; and
2. Re-enable AWS Shield auto-mitigation when it gets removed by a Terraform apply action.

Solution:
https://github.com/humanmade/shield-custom-aws-config-rule

# Related
- hashicorp/terraform-provider-aws#22869

**Comment by patheard** [comment]
This should now be fixed as of the AWS provider 4.56 release:
- hashicorp/terraform-provider-aws#29575

**Comment by patheard** [comment]
Currently the auto-mitigation still needs to be manually enabled until this lands:

- hashicorp/terraform-provider-aws#26873

**Comment by patheard** [comment]
Confirmed that auto-mitigation rule is now present in the test account and [TF is not attempting to manage it](https://github.com/cds-snc/url-shortener/actions/runs/5027639656/jobs/9017372373):
![image](https://github.com/cds-snc/scan-files/assets/2110107/813e30f2-94af-4ecc-9fbb-a4d84e760ddf)

**Comment by sakisv** [comment]
Hi folks,

I'm an SRE with the GOV.UK Notify team and we have some issues with the ShieldMitigationRuleGroup and I'd like to double check that it's not just us.

### tl;dr

With the terraform-aws-provider version 4.56.0, when we do a `plan` it seems to ignore any `ShieldMitigationRuleGroup` as expected. However, after we do an `apply` and we check the AWS Console, that `ShieldMitigationRuleGroup` is no longer there. 

### Example

We have 3 rules:
- The first is managed by terraform, and we're just changing the threshold
- The second is not managed by terraform and should be deleted on apply
- The third is a rule with a name that matches the pattern that the `wafv2_web_acl` ignores, so it should not appear in the plan and should not be removed

![Screenshot 2023-07-05 at 18 34 19](https://github.com/cds-snc/scan-files/assets/4289633/6dc93cfe-9fbd-49d8-8f36-f8f7e47a46f0)

Terraform plan shows the expected output:

```hcl
 # module.paas_proxy.aws_wafv2_web_acl.api_paas_cloudfront will be updated in-place
  ~ resource "aws_wafv2_web_acl" "api_paas_cloudfront" {
        id         = "633d1518-f001-4e32-a962-de935dc29e12"
        name       = "apiWaf"
        tags       = {}
        # (5 unchanged attributes hidden)

      - rule {
          - name     = "2000-Requests-per-second" -> null
          - priority = 0 -> null

          - action {

              - count {
                }
            }

          - statement {

              - rate_based_statement {
                  - aggregate_key_type = "IP" -> null
                  - limit              = 600000 -> null
                }
            }

          - visibility_config {
              - cloudwatch_metrics_enabled = true -> null
              - metric_name                = "BlockedOnMaxRequestsPerIP" -> null
              - sampled_requests_enabled   = true -> null
            }
        }
      + rule {
          + name     = "2000-Requests-per-second"
          + priority = 1

          + action {

              + count {
                }
            }

          + statement {

              + rate_based_statement {
                  + aggregate_key_type = "IP"
                  + limit              = 2000
                }
            }

          + visibility_config {
              + cloudwatch_metrics_enabled = true
              + metric_name                = "BlockedOnMaxRequestsPerIP"
              + sampled_requests_enabled   = true
            }
        }
      - rule {
          - name     = "another-rule-test" -> null
          - priority = 1 -> null

          - action {

              - count {
                }
            }

          - statement {

              - byte_match_statement {
                  - positional_constraint = "EXACTLY" -> null
                  - search_string         = "this-is-a-test" -> null

                  - field_to_match {

                      - single_header {
                          - name = "x-another-waf-test" -> null
                        }
                    }

                  - text_transformation {
                      - priority = 0 -> null
                      - type     = "NONE" -> null
                    }
                }
            }

          - visibility_config {
              - cloudwatch_metrics_enabled = true -> null
              - metric_name                = "another-rule-test" -> null
              - sampled_requests_enabled   = true -> null
            }
        }

        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

But after the apply this is what we see in the console:

![Screenshot 2023-07-06 at 12 15 31](https://github.com/cds-snc/scan-files/assets/4289633/6a580df0-9079-4a1d-9488-6c44a4534990)

Have you noticed a similar behaviour?

**Comment by patheard** [comment]
That is odd.  I did a quick check and I'm still seeing `ShieldMitigationRuleGroup` for accounts that have had a Terraform apply run since we enrolled.  However, none of those TF applies were making WAF ACL rule changes.

I'm off work this week, but when I'm back I'll run some tests to see if I can replicate what you're seeing.

**Comment by sakisv** [comment]
I double checked and it looks that the `ShieldMitigationRuleGroup` is being removed even if you change just the threshold for one of the rules that you manage in tf.

**Comment by sakisv** [comment]
That's great, thanks for confirming 👍
